source/Examples/SmartCache/README.md

```diff
@@ -0,0 +1,123 @@
+
+    ```shell
+    docker run -it -v $PWD:/src/MLOS --name mlos-build mlos/build:ubuntu-20.0
+    ```
+
+3. Inside the container, [build the compiled software](../../../documentation/02-Build.md#cli-make) with `make`:
+
+    ```sh
+    make dotnet-build cmake-build
+    ```
+
+    > This will build everything using a default `CONFIGURATION=Release`.
+    >
+    > To just build `SmartCache` and `Mlos.Agent.Server`, execute the following: \
+    > `make -C source/Examples/SmartCache && make -C source/Mlos.Agent.Server`
+
+4. For a `Release` build, the relevant output will be at:
+
+    - `Mlos.Agent.Server`:
+
+        `out/dotnet/source/Mlos.Agent.Server/obj/AnyCPU/Mlos.Agent.Server.dll`
+
+    - `SmartCache`:
+
+        `out/cmake/Release/source/Examples/SmartCache/SmartCache`
+
+    - `SmartCache.SettingsRegistry`
+
+        `out/dotnet/source/Examples/SmartCache/SmartCache.SettingsRegistry/obj/AnyCPU/SmartCache.SettingsRegistry.dll`
+
+## Executing
+
+`SmartCache` can be invoked separate, or by the `Mlos.Agent.Server` itself.
+
+Once started, `SmartCache` will attempt to register its component specific set of shared memory messages with the `Mlos.Agent` in the `Mlos.Agent.Server` using some `Mlos.Core` component registration messages.  That message includes the name of the `SettingsRegistry` assembly (`.dll`) corresponding to that component's settings/messages.
+
+The `Mlos.Agent.Server` needs to be told where it can find those assemblies.  To do that we provide an `MLOS_SETTINGS_REGISTRY_PATH` environment variable.
+
+In this case we populate it with the path to the `SmartCache.SettingsRegistry.dll`:
+
+```sh
+export MLOS_SETTINGS_REGISTRY_PATH="out/dotnet/source/Examples/SmartCache/SmartCache.SettingsRegistry/obj/AnyCPU:$MLOS_SETTINGS_REGISTRY_PATH"
+```
+
+Next, we can the `Mlos.Server.Agent` using the `dotnet` command:
+
+```sh
+tools/bin/dotnet out/dotnet/source/Mlos.Agent.Server/obj/AnyCPU/Mlos.Agent.Server.dll
+# FIXME: This is missing the .json file to connect to the optimizer service.
+```
+
+The `Mlos.Agent` that gets started will then wait for a signal that the other side (`SmartCache`) has connected to the shared memory region before started to poll it for messages to process.
+
+To start the `SmartCache` process we first need another shell instance in the docker container:
+
+```sh
+docker exec -it mlos-build /bin/bash
+```
+
+Now, we can start `SmartCache` as follows:
+
+```sh
+out/cmake/Release/source/Examples/SmartCache/SmartCache
+```
+
+> To have `Mlos.Agent.Server` start `SmartCache` without having to start another shell in the docker container instance, add the path to the `SmartCache` binary as an argument to the `dotnet ... Mlos.Agent.Server` invocation.
+
+## Caveats
+
+- The system currently only supports one shared memory region and doesn't cleanup the shared memory after itself.
+
+    To help with this, we currently provide a helper script to remove previous incarnations of the shared memory regions:
```

Comment:
all of this is within the docker (might be obvious?)

Reply:
Yeah, I can make a comment about that

Reply:
Added a note at the bottom.  Works?